crates/conjure_core/src/rule_engine/rule.rs

```diff
@@ -84,43 +83,25 @@ impl Reduction {
     pub fn with_symbols(new_expression: Expression, symbols: SymbolTable) -> Self {
         Self {
             new_expression,
-            new_top: Expression::And(Metadata::new(), Vec::new()),
+            new_top: vec![],
             symbols,
         }
     }
 
     /// Represents a reduction that also adds a top-level constraint to the model.
-    pub fn with_top(new_expression: Expression, new_top: Expression) -> Self {
+    pub fn with_top(new_expression: Expression, new_top: Vec<Expression>) -> Self {
         Self {
             new_expression,
             new_top,
             symbols: SymbolTable::new(),
         }
     }
 
-    // Apply side-effects (e.g. symbol table updates
-    pub fn apply(self, model: &mut Model) {
-        model.extend_sym_table(self.symbols);
-
-        // TODO: (yb33) Remove it when we change constraints to a vector
-        if let Expression::And(_, exprs) = &self.new_top {
-            if exprs.is_empty() {
-                model.constraints = self.new_expression.clone();
-                return;
-            }
-        }
-
-        model.constraints = match self.new_expression {
-            Expression::And(metadata, mut exprs) => {
-                // Avoid creating a nested conjunction
-                exprs.push(self.new_top.clone());
-                Expression::And(metadata.clone_dirty(), exprs)
-            }
-            _ => Expression::And(
-                Metadata::new(),
-                vec![self.new_expression.clone(), self.new_top],
-            ),
-        };
+    // Apply side-effects (e.g. symbol table updates)
+    pub fn apply(self, model: &mut Model, constraint_idx: usize) {
+        model.extend_sym_table(self.symbols); // Add new assignments to the symbol table
+        model.constraints.extend(self.new_top.clone());
+        model.constraints[constraint_idx] = self.new_expression.clone();
```

Comment:
I think this should just apply the side effects, not handle adding the new expression to the model. 

 In my opinion, adding the expression back into the model is rewriter-implementation specific, and in full generality you can't assume that the new expression is top level.

```suggestion
    // Apply side-effects (e.g. symbol table updates)
    pub fn apply(self, model: &mut Model) {
        model.extend_sym_table(self.symbols); // Add new assignments to the symbol table
        model.constraints.extend(self.new_top.clone());
```

Reply:
(For example,  I add expressions back into the tree in a slightly different way in #528, and having new_expression in apply caused it to break ( #527)

Reply:
I'm a little bit busy right now as I am preparing for the talk tommorow and revising, but I'll be happy to look at your suggestion a little bit later!

Reply:
No worries; I know the semester is officially over, so no rush!

I'm hopeful that, with these small changes, this could be merged, which would be a nice way to end the year!

Reply:
Agree with this change.